### DISCUSSION TITLE
How to access connection params from apollo client in context function?

### DISCUSSION THREAD

**Discussion by capaj** [mercurius-js/mercurius#981] [Q&A]
https://github.com/mercurius-js/mercurius/blob/master/docs/subscriptions.md#build-a-custom-graphql-context-object-for-subscriptions

there is an example given here:

![image](https://user-images.githubusercontent.com/1305378/232463538-687cff99-52bd-48f7-aab4-7a10bfac899f.png)

how do we get `Authorization` header?
My FE is using apollo client and I define it like this:
```ts
  const wsLink = new GraphQLWsLink(
    createClient({
      shouldRetry: () => true,
      retryAttempts: 20,
      lazy: true,
      url: graphqlApiUrl,
      connectionParams: async () => {
        const token = await getToken();
        return { Authorization: token };
      },
      keepAlive: 10000,
    })
  );
``` 

the problem is I am only able to see `Authorization` in my `onConnect` data.payload.
I am not able to access it through headers in context.
It would be nice if I could pass that payload into context somehow. Alternatively if I could access connection in the onConnect I could identify it and associate it on my side I think.

**Comment by capaj** [comment]
Oh it is copied into ctx.request.headers inside a resolver, just not inside the context function. 

```
{
  origin: 'http://10.0.2.2:3000',
  'sec-websocket-protocol': 'graphql-transport-ws',
  upgrade: 'websocket',
  connection: 'Upgrade',
  'sec-websocket-key': 'wHPw39tWDHVQgESFvYsI+A==',
  'sec-websocket-version': '13',
  'sec-websocket-extensions': 'permessage-deflate',
  host: '10.0.2.2:3000',
  'accept-encoding': 'gzip',
  'user-agent': 'okhttp/4.9.3',
  Authorization: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJhd3NfdXNlcl9pZCI6ImRkMjUzYmNkLTQ1MGMtNDgxYS05MWNhLTM2OTljMzM1MGZjZiIsInNlc3Npb25JRCI6IjdlZjMyMzhmLTE4MzItNDMwZS04NDNkLTY2ZTlkZGY3NDg4MSIsImlhdCI6MTY4MTcyNDc0OCwiZXhwIjoxNjgyOTM0MzQ4fQ.kcEqnY14LCkJPPn_oq3-7Kqe46ceLctYkDl8uUPvXfk'
}


```

 Why would it not be available in `context` request.headers param ?

**Reply by mcollina** [reply]
Because those headers are received with the first websocket message, we just add them in after the fact.
The context callback is processed before that happens.

**Reply by capaj** [reply]
thank @mcollina I thought that would be it, I saw the log from onConnect fire after the context callback, I just did not know how to formulate it in writing.
Is there no way to hold off on context callback untill after onConnect?

**Reply by mcollina** [reply]
Not in the current codebase, but I might be wrong and a scavenge in the code might be possible.
